src/audio/smart_amp_aw/aw_smart_amp.c

```diff
@@ -0,0 +1,557 @@
+//
+// Author: Jimmy Zhang <zhangjianming@awinic.com>
+
+#include <rtos/init.h>
+#include <sof/trace/trace.h>
+#include <sof/ipc/msg.h>
+#include <sof/ut.h>
+#include <user/smart_amp.h>
+#include <sof/audio/ipc-config.h>
+#include <sof/audio/smart_amp/aw_smart_amp.h>
+
+static const struct comp_driver comp_smart_amp;
+
+/* 0cd84e80-ebd3-11ea-adc1-0242ac120002 */
+DECLARE_SOF_RT_UUID("Awinic SKTune", awinic_sktune_comp_uuid, 0x0cd84e80, 0xebd3,
+		    0x11ea, 0xad, 0xc1, 0x02, 0x42, 0xac, 0x12, 0x00, 0x02);
+
+DECLARE_TR_CTX(awinic_sktune_comp_tr, SOF_UUID(awinic_sktune_comp_uuid),
+	       LOG_LEVEL_INFO);
+
+/* Amp configuration & model calibration data for tuning/debug */
+#define SOF_SMART_AMP_CONFIG 0
+#define SOF_SMART_AMP_MODEL 1
+
+
+struct smart_amp_data {
+	struct sof_smart_amp_config config;
+	struct comp_buffer *source_buf; /**< stream source buffer */
+	struct comp_buffer *feedback_buf; /**< feedback source buffer */
+	struct comp_buffer *sink_buf; /**< sink buffer */
+	uint32_t in_channels;
+	uint32_t out_channels;
+	struct ipc_config_process ipc_config;
+	/* module handle for speaker protection algorithm */
+	sktune_t *algo_handle;
+};
+
+static struct comp_dev *smart_amp_comp_new(const struct comp_driver *drv,
+				      const struct comp_ipc_config *config,
+				      const void *spec)
+{
+	struct comp_dev *dev;
+	const struct ipc_config_process *ipc_cfg = spec;
+	struct smart_amp_data *amp_data;
+	struct sof_smart_amp_config *cfg;
+	size_t block_size;	
+	int ret;
+
+	dev = comp_alloc(drv, sizeof(*dev));
+	if (!dev)
+		return NULL;
+
+	dev->ipc_config = *config;
+
+	amp_data = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*amp_data));
+	if (!amp_data) {
+		comp_err(dev, "amp_data alloc failed");
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, amp_data);
+	amp_data->ipc_config = *ipc_cfg;
+
+	cfg = (struct sof_smart_amp_config *)ipc_cfg->data;  
+	block_size = ipc_cfg->size;
+
+	if (block_size > 0 && block_size < sizeof(struct sof_smart_amp_config)) {
+		comp_err(dev, "smart_amp_new(): failed to apply config");
+		goto error;
+	}
+
+	memcpy_s(&amp_data->config, sizeof(struct sof_smart_amp_config), cfg, block_size);
+
+	/*memory allocation for sktune*/
+	amp_data->algo_handle = smart_amp_sktune_alloc(dev);
+	if (!amp_data->algo_handle) {
+		comp_err(dev, "[Awinic] SKTune alloc failed!");
+		goto error;
+	}
+
+
+	/* Bitwidth information is not available. Use 16bit as default.*/
+	amp_data->algo_handle->media_info.bit_per_sample = 16;
+	amp_data->algo_handle->media_info.bit_qactor_sample = 15;
+	amp_data->algo_handle->media_info.num_channel = 2;
+	amp_data->algo_handle->media_info.sample_rate = 48000;
+	amp_data->algo_handle->media_info.data_is_signed = 1;
+
+	ret = smart_amp_init(amp_data->algo_handle, dev);
+	if (ret) {
+		comp_err(dev, "[Awinic] samrt amp init failed!");
+		goto error;
+	}
+
+	dev->state = COMP_STATE_READY;
+
+	return dev;
+
+error:
+	rfree(amp_data->algo_handle);
```

Comment:
Hi Awinic folks,

An important note for dynamic memory allocation in SOF 3P mpdule design: you should always keep the transparency for dynamic memory occupancy. That is, NEVER allocate dynamic memory inside the library directly through stdio functions e.g. malloc. There are two alternative ways:

1. explicitly define the buffer you'll need inside the library, and do the allocation in external domain on module initialization (same for de-allocation on module release)

    Example code for Maxim smart_amp, which defines all struct objects in external domain, and do memory allocation/de-allocation at once:
    `smart_amp_alloc_memory()`
    https://github.com/thesofproject/sof/blob/main/src/audio/smart_amp/smart_amp.c#L76

    `smart_amp_free_memory()`
    https://github.com/thesofproject/sof/blob/main/src/audio/smart_amp/smart_amp.c#L47

    Example code for Waves in module_adapter, it preserves the struct definition, an API function is built to 
query the memory size needed inside from external domain `MaxxEffect_GetMessageMaxSize()`
https://github.com/thesofproject/sof/blob/main/src/include/sof/audio/MaxxEffect/Control/RPC/MaxxEffect_RPC_Server.h#L26

    Obtain the required size and allocate the memory  in external domain,
https://github.com/thesofproject/sof/blob/main/src/audio/module_adapter/module/waves.c#L194

3. if you really need the inline memory allocation (allocate the memory in internal domain), you will need to create the specified API to expose allocate/free to external domain, and use them to allocate memory in internal

    Example code for DTS in module_adapter:
    `dts_effect_allocate_codec_memory()`
https://github.com/thesofproject/sof/blob/main/src/audio/module_adapter/module/dts.c#L19

    `dts_effect_allocate_free_memory()`
https://github.com/thesofproject/sof/blob/main/src/audio/module_adapter/module/dts.c#L38

    The implementation requires an external structure for monitoring, e.g. `struct module_memory`. By this method the external domain will have counterpart of the set of allocated buffer pointers
https://github.com/thesofproject/sof/blob/main/src/include/sof/audio/module_adapter/module/generic.h#L131

Last but not least, for data buffer allocation, always use the following function to make sure the correct memory region is applied:  
```
memory = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, size);
```

Thanks

Reply:
> for data buffer allocation, always use the following function to make sure the correct memory region is applied:
> 
> ```
> memory = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, size);
> ```

I think for buffer allocations `rballoc()` or `rballoc_align()` should be used

Reply:
ah yes @lyakh is correct, please use the following command instead (for allocating data buffers):
```
memory = rballoc(0, SOF_MEM_CAPS_RAM, size);
```
Please note the memory allocated from rballoc() won't be preset to all-zeros since most use cases for data buffers don't need that. If needed, please use bzero() after that.